### ISSUE TITLE
Inference of latent variables of new points in GPLVM

### ISSUE THREAD

**Issue by mrksr** [GPflow/GPflow#286]
Hi everyone,

I was very happy to see that GPLVMs are finding their way into GPflow, which is very interesting to me. Since I would like to use them in a prediction-setting, I am wondering how they can be used to infer the latent variables of an unseen observed point `Ynew`, analogous to the `infer_newX`-method in [GPy](https://github.com/SheffieldML/GPy/blob/devel/GPy/inference/latent_function_inference/inferenceX.py). Additionally, I would love for this inference to support missing values in this new sample.

Is this already in place in GPflow? If not, what would be the correct way of implementing it? I would be happy to help.

Thank you for all your work!

**Comment by markvdw** [comment]
None of this is currently supported. I'm not sure about missing data, but a bit of a hack for `infer_newX` could be to just add a new observation and initialisation to the `Y`, `X_mean` and `X_var` matrices, and then optimise. I'm not sure what GPy does, but this would technically be the correct thing to do.

**Comment by alexisboukouvalas** [comment]
thanks @mrksr , glad you are finding the GPLVM support interesting!
As @markvdw says we currently do not support either of these features (missing data, inference on X). We have an issue currently ( #274)  with suggestions on how to improve the latent variable model code. 
Any help appreciated!

**Comment by mrksr** [comment]
I think it might be a good idea to start with the inference without missing values. I hope the generalization to missing values will not be too hard then. If I understand the code in GPy correctly, they are essentially cloning the model and setting up the optimization @markvdw described.

Would the same thing be the way to go with GPflow, i.e. cloning the model, fixing all existing parameters and adding parameters for the new points? I think it would be more correct to optimize all parameters and integrate this into the original model, but this would probably take way too much time for every prediction. Plus, it might not really be desirable once there is missing data.

**Comment by mrksr** [comment]
Just a ping to @alexisboukouvalas and @jameshensman, could you maybe give a hint as to how to implement the inference nicely into GPflow?

I think in a more general sense, the problem of which this inference is an instance of is that given some already existing model, one would have to formulate a derived one which shares some parameters (and maybe fixes them) but also adds a few new ones.

**Comment by jameshensman** [comment]
Hmm, making a nice interface might be a bit tricky. 

I would add some new methods to the (B)GPLVM class. First, an autoflow method that computes the objective function given (Ynew, mu_new and log_S_new). Then, another method that returns a function suitable for scipy.minimize, and finally a convenience function with sensible defaults for doing the optimization. 

Sketch:
```python
class BayesianGPLVM(...):
    <snip>

    @AutoFlow(...)
    def held_out_data_objective(self, Ynew, mu_new, log_S_new):
        # your code here
        return objective, gradients

    def _held_out_data_wrapper_creator(self, Ynew):
        def fun(x_flat):
            mu = x_flat[::2].reshape(??)
            logvar = x_flat[1::2].reshape(??)
            return self.held_out_data_objective(Ynew, mu_new, log_S_new)
        return fun

    def infer_latent_inputs(self, Ynew):
        f = self._held_out_data_wrapper_creator(Ynew)
        x_init = np.array(...)  # something sensible
        x_hat = scipy.optimize.minimize(f, x_init, other_sensible_defaults)
        mu, var = x_hat[::2].reshape(??), x_hat[1::2].reshape(??)
        return mu, var
```

This design could probably be refined, but suggest you get this working to start.

**Comment by mrksr** [comment]
Thanks! I will report back once I have some results.

**Comment by markvdw** [comment]
Also, missing data can be handled by making the likelihood precision zero, right? This shouldn't be too hard to add I think

**Comment by javdrher** [comment]
Yesterday I quickly implemented inversion for standard GP models. 
I did it by fixing all Params, introducing a new Param for Xnew and adding Ynew. When the Xnew param is present when tf_mode is active I concat it with X. Doing so I could use optimize() to obtain Xnew. Here's the code https://github.com/javdrher/GPflow/blob/master/GPflow/model.py

The principle is applicable for Bayesian GP-LVMs too I guess with some slight differences (iirc the KL term for the inversion only considers Xnew?), but I'm unsure if its a better approach than your suggestion. I specifically don't like the property I introduced (done mostly due to lack of time), a better way would be a method `_get_free_X()` or something alike.
